src/deploy/functions/cache/hash.ts

```diff
@@ -0,0 +1,69 @@
+import { readFile } from "node:fs/promises";
+import * as crypto from "crypto";
+import { Backend, Endpoint } from "../backend";
+import { getSecretVersions } from "../../../functions/secrets";
+
+/**
+ * Generates a hash from the environment variables of a {@link Backend}.
+ * @param backend Backend of a set of functions
+ */
+export function getEnvironmentVariablesHash(backend: Backend): string {
+  const hash = crypto.createHash("sha256");
+
+  // Hash the contents of the dotenv variables
+  const hasEnvironmentVariables = !!Object.keys(backend.environmentVariables).length;
+  if (hasEnvironmentVariables) {
+    hash.update(JSON.stringify(backend.environmentVariables));
+  }
+
+  return hash.digest("hex");
+}
+
+/**
+ * Retrieves the unique hash given a pathToGeneratedPackageFile.
+ * @param pathToGeneratedPackageFile Packaged file contents of functions
+ */
+export async function getSourceHash(pathToGeneratedPackageFile?: string): Promise<string> {
```

Comment:
q: When is `pathToGeneratedPackageFile` undefined during function deploy?

Reply:
Good point! It shouldn't be! Updated!